R/utils.R

```diff
@@ -437,18 +442,122 @@ rot_x <- function(angle=30, hjust = 1, vjust = NULL, vertical = FALSE, ...) {
     }
     angle <- 90
     vjust <- 0.5
+  } else {
+    if(is.character(hjust)) {
+      abort("hjust must be numeric or NULL in this case.")
+    }
   }
   x <- element_text(angle = angle, hjust = hjust, vjust = vjust, ...)
   theme(axis.text.x=x)
 }
 
 ##' @rdname rot_x
 ##' @export
-rot_y <- function(angle=30, hjust = 1, vjust = NULL,...) {
-  y <- element_text(angle = angle, hjust = hjust, vjust = vjust, ...)
+rot_y <- function(angle=30, hjust = 1, vjust = NULL, vertical = FALSE, ...) {
+  if(isTRUE(vertical)) {
+    if(is.character(vjust)) {
+      vjust <- match.arg(vjust, c("left", "right"))
+      if(vjust=="left") vjust <- 1
+      if(vjust=="right") vjust <- 0
+    } else {
+      if(missing(vjust)) vjust <- 1
+    }
+    angle <- 90
+    hjust <- 0.5
+  } else {
+    if(is.character(vjust)) {
+      abort("`vjust` must be numeric or NULL in this case.")
+    }
+  }
+  y <- element_text(angle = angle, hjust = hjust, vjust = vjust,, ...)
   theme(axis.text.y=y)
 }
 
+
+.rotxy <- function(x, ...) UseMethod(".rotxy")
+#' @export
+.rotxy.gg <- function(x, axis = "x", ...) {
+  if(axis=="x") {
+    x <- x + rot_x(...)
+  } else {
+    x <- x + rot_y(...)
+  }
+  x
+}
+#' @export
+.rotxy.patchwork <- function(x, axis = "x", ...) {
+  stopifnot(requireNamespace("patchwork", quietly = TRUE))
+  if(axis=="x") {
+    x <- x & rot_x(...)
+  } else {
+    x <- x & rot_y(...)
+  }
+  x
+}
+#' @export
+.rotxy.default <- function(x, ...) {
```

Comment:
Do we need this? Maybe just let dispatch fail if the object isn't the right class.

Reply:
If I'm tracing things right, the only way for the user to access `.rotxy.default` is via `rot_at`.

```R
x <- list(1, 2)
names(x) <- c("one", "two")

# without .default
rot_at(x)
#> Error in .rotxy.default(x[[w]], axis = axis, ...) :
#>  invalid object; can only rotate gg or patchwork objects.

# without .default
rot_at(x)
#> Error in UseMethod(".rotxy") :
#>   no applicable method for '.rotxy' applied to an object of class "c('integer', 'numeric')"
```

I don't see any issues with the `.default` variant, though in my opinion the vanilla message is good enough here.

Reply:
Thanks; I'll remove the default. The no-default message is probably better than what I came up with.